espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ViewFinder.kt

```diff
@@ -125,72 +125,72 @@ object ViewFinder {
             Strategy.TEXT ->
                 // with text
-                views = getViews(root, withText(selector), findOne)
+                return getViews(root, withText(selector), findOne).map { ViewState(it) }
             Strategy.ACCESSIBILITY_ID -> {
-                views = getViews(root, withContentDescription(selector), findOne)
+                val result = getViews(root, withContentDescription(selector), findOne)
 
                 // If the item is not found on the screen, use 'onData' to try
                 // to scroll it into view and then locate it again
-                if (views.isEmpty() && canScrollToViewWithContentDescription(root, selector)) {
-                    views = getViews(root, withContentDescription(selector), findOne)
-                }
+                return if (result.isEmpty() && canScrollToViewWithContentDescription(root, selector))
+                    getViews(root, withContentDescription(selector), findOne).map { ViewState(it) }
+                else
+                    result.map { ViewState(it) }
             }
             Strategy.XPATH ->
                 // If we're only looking for one item that matches xpath, pass it index 0 or else
                 // Espresso throws an AmbiguousMatcherException
-                views = if (findOne) {
-                    getViews(root, withXPath(root, selector, 0), true)
+                return if (findOne) {
+                    getViews(root, withXPath(root, selector, 0), true).map { ViewState(it) }
                 } else {
-                    getViews(root, withXPath(root, selector), false)
+                    getViews(root, withXPath(root, selector), false).map { ViewState(it) }
                 }
-            Strategy.VIEW_TAG -> views = getViews(root, withTagValue(allOf(instanceOf(String::class.java),
-                    equalTo(selector as Any))), findOne)
+            Strategy.VIEW_TAG ->
+                return getViews(root, withTagValue(allOf(instanceOf(String::class.java), equalTo(selector as Any))), findOne)
+                    .map { ViewState(it) }
             Strategy.DATAMATCHER -> {
                 val matcher = selector.toJsonMatcher()
-                views = try {
+                return try {
                     getViewsFromDataInteraction(root, onData(matcher.query.matcher))
+                        .map { ViewState(it) }
                 } catch (e: PerformException) {
                     // Perform Exception means nothing was found. Return empty list
                     emptyList()
                 }
             }
             Strategy.VIEWMATCHER -> {
                 val matcherJson = selector.toJsonMatcher()
-                views = try {
+                return try {
                     @Suppress("UNCHECKED_CAST")
                     getViews(root, matcherJson.query.matcher as Matcher<View>, findOne, matcherJson.query.scope as Matcher<Root>)
```

Comment:
Shouldn't this scope be nullable?

Reply:
it is nullable. Does something looks suspicious here?

Reply:
If I not mistaken the cast should be Matcher<Root>?, when the scope is null it will throw an Exception

Reply:
It should work just fine, `as?` is needed to do cast from types other than null: https://stackoverflow.com/questions/50836017/safe-cast-vs-cast-to-nullable